man/mce.8
.TH MCE 8 "Jan 20, 2011" "Nokia"

.SH NAME
mce \- daemon that provides mode control functionality

.SH SYNOPSIS
.B mce
[\fIOPTION\fP]...

.SH DESCRIPTION
.B mce
is a daemon that provides mode control functionality.

On startup,
.B mce
will read the initial states of the various radios from the
radio states file and broadcast these states.

Notifiers are registered for various GConf entries,
as well as
\%D\(hyBus message handlers for the method calls that
.B mce
provides, and \%D\(hyBus signal handlers for the signals that
.B mce
listens to.

.B mce
will also listen to key events from the power button.
If the user preses the power button, the action configured in
.I /etc/mce/mce.ini
will be executed.

Among the other functionality handled by
.B mce
is the logic for the touch screen and keypad lock,
LED patterns and display backlight control.

.SH OPTIONS
.TP
.BR \-d , \ \-\-daemonflag
Run
.B mce
as a daemon
.TP
.B \-\-force\-syslog
Log to syslog even when mce isn't run as a daemon
.TP
.B \-\-force\-stderr
Log to stderr even when mce is run as a daemon
.TP
.BR \-S , \ \-\-session
Use the session bus instead of the system bus for \%D\(hyBus
communication
.TP
.B \-\-show\-module\-info
Show information about loaded modules
.TP
.B \-\-debug\-mode
Start mce even if communication with dsme fails
.TP
.B \-\-quiet
Decrease debug message verbosity
.TP
.B \-\-verbose
Increase debug message verbosity
.TP
.B \-\-help
Display help for the command
.TP
.B \-\-version
Display version and author information

.SH GCONF KEYS
.TP
.I /system/osso/dsm/dsm/locks/touchscreen_keypad_autolock_enabled
.B bool
(read only)
.br
If this key is set to
.BR true ,
the touchscreen/keypad lock
will automatically be enabled when the screen blanks.
If this key is set to
.BR false ,
the lock will only be activated if the user triggers it
manually.

.TP
.I /system/osso/dsm/display/display_brightness
.B int
(read only)
.br
The brightness profile to use for the screen.
Valid values range from
.B 1
to
.BR 5 .

.TP
.I /system/osso/dsm/display/display_dim_timeout
.B int
(read only)
.br
Number of seconds to wait before the screen is dimmed.

.TP
.I /system/osso/dsm/display/display_blank_timeout
.B int
(read only)
.br
Number of seconds to wait before the screen is blanked.

.SH FILES
.TP
.I /var/run/mce.pid
PID\(hyfile used by
.BR mce .
.TP
.I /var/run/mce/boot
File used by
.B mce
to detect whether the program was started on bootup or
restarted due to a crash/manual restart.
.TP
.I /var/lib/mce/radio_states
Stores the system radio states.
This file should never be edited manually,
and shall not be modified by any other process than
.BR mce ,
since the file is only read when
.B mce
is started.

.SH SEE ALSO
.BR mcetool (8)

.SH HISTORY
Jun 16 2010: Updated for new and modified functionality.
.br
Jun 13 2010: Various improvements.
.br
May 13 2010: Updated for modified functionality.
.br
Oct 19 2007: Use escape\(hycodes for hyphens.
.br
Apr 12 2007: Minor fixes.
.br
Oct 23 2006: Minor fixes.
.br
Oct 3 2006: Fix typo.
.br
Sep 23 2005: Fixed minor typos.
.br
Sep 21 2005: Initial version of this manual page.

.SH AUTHOR
mce and its manual page is written by
David Weinehall <\fIdavid.weinehall@nokia.com\fP>

.SH REPORTING BUGS
Report bugs to
<\fIdavid.weinehall@nokia.com\fP>.

.SH COPYRIGHT
Copyright \(co 2005\(hy2011 Nokia Corporation.  All rights reserved.


debian/changelog
mce (1.10.121) UNRELEASED; urgency=low

  * Ignore external events in proximity states (Fixes: NB#250798)

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Thu, 19 May 2011 11:21:00 +0300

mce (1.10.120) unstable; urgency=low

  * Add delay to brightness stepdown (Fixes: NB#253917)

 -- Jukka Turunen <ext-jukka.t.turunen@nokia.com>  Tue, 17 May 2011 13:22:03 +0300

mce (1.10.119) unstable; urgency=low

  * Request dsme::DeviceStateControl capability (Fixes: NB#255382)

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Thu, 12 May 2011 15:06:16 +0300

mce (1.10.118) unstable; urgency=low

  * Close LPM tklock on unlock request. (Fixes: NB#254339)
    | Thanks to Jukka Turunen

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Wed, 11 May 2011 08:37:45 +0300

mce (1.10.117) unstable; urgency=low

  * Unset low power mode gconf key on device clean

 -- Jukka Turunen <ext-jukka.t.turunen@nokia.com>  Mon, 09 May 2011 14:21:43 +0300

mce (1.10.116) unstable; urgency=low

  * Low power mode (screen saver)  enabled by default (Fixes: NB#247527)

 -- Jukka Turunen <ext-jukka.t.turunen@nokia.com>  Mon, 09 May 2011 12:45:48 +0300

mce (1.10.115) unstable; urgency=low

  * Fix for proximity locking when low power mode is enabled. Display
    now goes off instead of LPM when voice call is active (headset in
    use) (Fixes: NB#249789)
    | Thanks to Irina Bezruk

 -- Jyrki Hämäläinen <ext-jyrki.pe.hamalainen@nokia.com>  Wed, 04 May 2011 15:09:04 +0300

mce (1.10.114) unstable; urgency=low

  * Fix acting dead alarm logic mistake (Fixes: NB#251528)

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Wed, 04 May 2011 15:39:22 +0300

mce (1.10.113+nmu3) unstable; urgency=low

  * Checking in modules/keypad.c to guard against division by
    zero (Fixes: NB#252430)

 -- Jukka Turunen <ext-jukka.t.turunen@nokia.com>  Wed, 04 May 2011 13:53:36 +0300

mce (1.10.113+nmu2) unstable; urgency=low

  * idle_interval values in ms, was Hz (Fixes: NB#240705)

 -- Jukka Turunen <ext-jukka.t.turunen@nokia.com>  Wed, 04 May 2011 11:21:22 +0300

mce (1.10.113+nmu1) unstable; urgency=low

  * Set tklock timer when call active, disable proximity locking with
    headset (Fixes: NB#248570)

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Mon, 02 May 2011 13:53:29 +0300

mce (1.10.113) unstable; urgency=low

  * Fixed incorrect logic in the submode trigger inside modules/display.c
  * Filter out requests to turn on the display when the device
    is in acting dead, unless there's an active alarm
    (Fixes: NB#237750)

 -- David Weinehall <david.weinehall@nokia.com>  Wed, 27 Apr 2011 23:29:13 +0300

mce (1.10.112) unstable; urgency=low

  * Load the ALS module before the display module, to ensure that the
    brightness filter are available before we attempt to filter the
    brightness settings (Fixes: NB#239982, NB#250357)
  * Modified the display blanking inhibit to also cover alarms
    (Fixes: NB#247420)
  * Don't disable touchscreen/keypad when alarm UI state changes to off
    if the visual tklock UI is active (Fixes: NB#248865)

 -- David Weinehall <david.weinehall@nokia.com>  Wed, 27 Apr 2011 05:02:05 +0300

mce (1.10.111) unstable; urgency=low

  * Do not autorelock when calls end (Fixes: NB#242930)
    | Thanks to Victor Portnov
  * Modified the proximity logic for LPM a bit

 -- David Weinehall <david.weinehall@nokia.com>  Tue, 26 Apr 2011 12:28:58 +0300

mce (1.10.110) unstable; urgency=low

  * Double tap gesture related proximity should only be processed
    when the display is off or in low power mode (Fixes: NB#249263)
  * Request sys_rawio and sys_admin credentials, needed due to stricter
    security for I2C

 -- David Weinehall <david.weinehall@nokia.com>  Thu, 21 Apr 2011 16:08:51 +0300

mce (1.10.109) unstable; urgency=low

  * Set ALS thresholds to [0, 65535] when the ALS isn't in use,
    to ensure that it does not trigger unnecessary wakeups
    (Fixes: NB#249176)

 -- David Weinehall <david.weinehall@nokia.com>  Wed, 20 Apr 2011 10:45:20 +0300

mce (1.10.108) unstable; urgency=low

  * Disable doubletap gesture when there's proximity
    (Fixes: NB#240705)
  * Unset the GConf key for the double tap gesture in mce-device-clear
    (Fixes: NB#245318)
    | Thanks to Victor Portnov
  * Handle corner cases with double tap gesture + proximity + call/alarm
    (Fixes: NB#240720)
    | Thanks to Alexadner Shertz
  * Inhibit proximity sensor during call if manually blanked
    (Fixes: NB#239703)
    | Thanks to Irina Bezruk
  * Added synthetic powerkey repeat events to tklock code,
    to keep the display from blanking when the user keeps the power
    button pressed for longer than the visual blank timeout
    (Fixes: NB#235482)
    | Thanks to Alexadner Shertz
  * Set proximity sensor to provide continuous data when the external
    reference count is > 0 (Fixes: NB#248161)
  * Changed default brightness setting to 3 instead of 4
    (Fixes: NB#248535)

 -- David Weinehall <david.weinehall@nokia.com>  Tue, 19 Apr 2011 13:52:22 +0300

mce (1.10.107) unstable; urgency=low

  * Added blank timeout for low power mode when there's proximity
    (Fixes: NB#246627)
  * Increase the idle interval for the touchscreen to minimise
    power consumption when the screen is off and lower it when waiting
    for the tklock swipe gesture
  * Modified the HBM use-time tracking to report correct values when
    the device doesn't support high brightness mode
  * Fixed two stupid mistakes in the activity callback code
    (Fixes: NB#239760)

 -- David Weinehall <david.weinehall@nokia.com>  Thu, 14 Apr 2011 13:56:51 +0300

mce (1.10.106) unstable; urgency=low

  * Added support for the activity callback API (Fixes: NB#239760)

mce.c
/**
 * @file mce.c
 * Mode Control Entity - main file
 * <p>
 * Copyright © 2004-2011 Nokia Corporation and/or its subsidiary(-ies).
 * <p>
 * @author David Weinehall <david.weinehall@nokia.com>
 *
 * mce is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License
 * version 2.1 as published by the Free Software Foundation.
 *
 * mce is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with mce.  If not, see <http://www.gnu.org/licenses/>.
 */
#include <glib.h>
#include <glib-object.h>		/* g_type_init() */

#include <errno.h>			/* errno, ENOMEM */
#include <fcntl.h>			/* open(), O_RDWR, O_CREAT */
#include <stdio.h>			/* fprintf(), sprintf(),
					 * stdout, stderr
					 */
#include <getopt.h>			/* getopt_long(),
					 * struct options
					 */
#include <signal.h>			/* signal(),
					 * SIGTSTP, SIGTTOU, SIGTTIN,
					 * SIGCHLD, SIGUSR1, SIGHUP,
					 * SIGTERM, SIG_IGN
					 */
#include <stdlib.h>			/* exit(), EXIT_FAILURE, EXIT_SUCCESS */
#include <string.h>			/* strlen() */
#include <unistd.h>			/* close(), lockf(), fork(), chdir(),
					 * getpid(), getppid(), setsid(),
					 * write(), getdtablesize(), dup(),
					 * F_TLOCK
					 */
#include <sys/stat.h>			/* umask() */

#include "mce.h"			/* _(),
					 * setlocale() -- indirect,
					 * bindtextdomain(),
					 * textdomain(),
					 * mainloop,
					 * system_state_pipe,
					 * master_radio_pipe,
					 * call_state_pipe,
					 * call_type_pipe,
					 * submode_pipe,
					 * display_state_pipe,
					 * display_brightness_pipe,
					 * led_brightness_pipe,
					 * led_pattern_activate_pipe,
					 * led_pattern_deactivate_pipe,
					 * key_backlight_pipe,
					 * keypress_pipe,
					 * touchscreen_pipe,
					 * device_inactive_pipe,
					 * lockkey_pipe,
					 * keyboard_slide_pipe,
					 * lid_cover_pipe,
					 * lens_cover_pipe,
					 * proximity_sensor_pipe,
					 * tk_lock_pipe,
					 * charger_state_pipe,
					 * battery_status_pipe,
					 * battery_level_pipe,
					 * inactivity_timeout_pipe,
					 * audio_route_pipe,
					 * usb_cable_pipe,
					 * jack_sense_pipe,
					 * power_saving_mode_pipe,
					 * thermal_state_pipe,
					 * MCE_STATE_UNDEF,
					 * MCE_INVALID_MODE_INT32,
					 * CALL_STATE_NONE,
					 * NORMAL_CALL,
					 * MCE_ALARM_UI_INVALID_INT32,
					 * MCE_NORMAL_SUBMODE,
					 * MCE_DISPLAY_UNDEF,
					 * LOCK_UNDEF,
					 * BATTERY_STATUS_UNDEF,
					 * THERMAL_STATE_UNDEF,
					 * DEFAULT_INACTIVITY_TIMEOUT
					 */

#include "mce-log.h"			/* mce_log_open(), mce_log_close(),
					 * mce_log_set_verbosity(), mce_log(),
					 * LL_*
					 */
#include "mce-conf.h"			/* mce_conf_init(),
					 * mce_conf_exit()
					 */
#include "mce-dbus.h"			/* mce_dbus_init(),
					 * mce_dbus_exit()
					 */
#include "mce-dsme.h"			/* mce_dsme_init(),
					 * mce_dsme_exit()
					 */
#include "mce-gconf.h"			/* mce_gconf_init(),
					 * mce_gconf_exit()
					 */
#include "mce-modules.h"		/* mce_modules_dump_info(),
					 * mce_modules_init(),
					 * mce_modules_exit()
					 */
#include "event-input.h"		/* mce_input_init(),
					 * mce_input_exit()
					 */
#include "event-switches.h"		/* mce_switches_init(),
					 * mce_switches_exit()
					 */
#include "datapipe.h"			/* setup_datapipe(),
					 * free_datapipe()
					 */
#include "modetransition.h"		/* mce_mode_init(),
					 * mce_mode_exit()
					 */

/* "TBD" Modules; eventually this should be handled differently */
#include "tklock.h"			/* mce_tklock_init(),
					 * mce_tklock_exit()
					 */
#include "powerkey.h"			/* mce_powerkey_init(),
					 * mce_powerkey_exit()
					 */

/** Path to the lockfile */
#define MCE_LOCKFILE			"/var/run/mce.pid"
/** Name shown by --help etc. */
#define PRG_NAME			"mce"

extern int optind;			/**< Used by getopt */
extern char *optarg;			/**< Used by getopt */

static const gchar *progname;	/**< Used to store the name of the program */

/**
 * Display usage information
 */
static void usage(void)
{
	fprintf(stdout,
		_("Usage: %s [OPTION]...\n"
		  "Mode Control Entity\n"
		  "\n"
		  "  -d, --daemonflag           run MCE as a daemon\n"
		  "      --force-syslog         log to syslog even when not "
		  "daemonized\n"
		  "      --force-stderr         log to stderr even when "
		  "daemonized\n"
		  "  -S, --session              use the session bus instead\n"
		  "                               of the "
		  "system bus for D-Bus\n"
		  "      --show-module-info     show information about "
		  "loaded modules\n"
		  "      --debug-mode           run even if dsme fails\n"
		  "      --quiet                decrease debug message "
		  "verbosity\n"
		  "      --verbose              increase debug message "
		  "verbosity\n"
		  "      --help                 display this help and exit\n"
		  "      --version              output version information "
		  "and exit\n"
		  "\n"
		  "Report bugs to <david.weinehall@nokia.com>\n"),
		progname);
}

/**
 * Display version information
 */
static void version(void)
{
	fprintf(stdout, _("%s v%s\n%s"),
		progname,
		G_STRINGIFY(PRG_VERSION),
		_("Written by David Weinehall.\n"
		  "\n"
		  "Copyright (C) 2004-2010 Nokia Corporation.  "
		  "All rights reserved.\n"));
}

/**
 * Initialise locale support
 *
 * @param name The program name to output in usage/version information
 * @return 0 on success, non-zero on failure
 */
static gint init_locales(const gchar *const name)
{
	gint status = 0;

#ifdef ENABLE_NLS
	setlocale(LC_ALL, "");

	if ((bindtextdomain(name, LOCALEDIR) == 0) && (errno == ENOMEM)) {
		status = errno;
		goto EXIT;
	}

	if ((textdomain(name) == 0) && (errno == ENOMEM)) {
		status = errno;
		goto EXIT;
	}

EXIT:
	/* In this error-message we don't use _(), since we don't
	 * know where the locales failed, and we probably won't
	 * get a reasonable result if we try to use them.
	 */
	if (status != 0) {
		fprintf(stderr,
			"%s: `%s' failed; %s. Aborting.\n",
			name, "init_locales", g_strerror(status));
	}

	if (errno != ENOMEM)
		errno = 0;
#endif /* ENABLE_NLS */
	progname = name;

	return status;
}

/**
 * Signal handler
 *
 * @param signr Signal type
 */
static void signal_handler(const gint signr)
{
	switch (signr) {
	case SIGUSR1:
		/* We'll probably want some way to communicate with MCE */
		break;

	case SIGHUP:
		/* Possibly for re-reading configuration? */
		break;

	case SIGTERM:
		/* This should be done through a pipe or signalfd instead */
		g_main_loop_quit(mainloop);
		break;

	default:
		/* Should never happen */
		break;
	}
}

/**
 * Daemonize the program
 *
 * @return TRUE if MCE is started during boot, FALSE otherwise
 */
static gboolean daemonize(void)
{
	gint retries = 0;
	gint i = 0;
	gchar str[10];

	if (getppid() == 1)
		goto EXIT;	/* Already daemonized */

	/* Detach from process group */
	switch (fork()) {
	case -1:
		/* Failure */
		mce_log(LL_CRIT, "daemonize: fork failed: %s",
			g_strerror(errno));
		mce_log_close();
		exit(EXIT_FAILURE);

	case 0:
		/* Child */
		break;

	default:
		/* Parent -- exit */
		exit(EXIT_SUCCESS);
	}

	/* Detach TTY */
	setsid();

	/* Close all file descriptors and redirect stdio to /dev/null */
	if ((i = getdtablesize()) == -1)
		i = 256;

	while (--i >= 0) {
		if (close(i) == -1) {
			if (retries > 10) {
				mce_log(LL_CRIT,
					"close() was interrupted more than "
					"10 times. Exiting.");
				mce_log_close();
				exit(EXIT_FAILURE);
			}

			if (errno == EINTR) {
				mce_log(LL_INFO,
					"close() was interrupted; retrying.");
				errno = 0;
				i++;
				retries++;
			} else if (errno == EBADF) {
				/* Ignore invalid file descriptors */
				errno = 0;
			} else {
				mce_log(LL_CRIT,
					"Failed to close() fd %d; %s. "
					"Exiting.",
					i + 1, g_strerror(errno));
				mce_log_close();
				exit(EXIT_FAILURE);
			}
		} else {
			retries = 0;
		}
	}

	if ((i = open("/dev/null", O_RDWR)) == -1) {
		mce_log(LL_CRIT,
			"Cannot open `/dev/null'; %s. Exiting.",
			g_strerror(errno));
		mce_log_close();
		exit(EXIT_FAILURE);
	}

	if ((dup(i) == -1)) {
		mce_log(LL_CRIT,
			"Failed to dup() `/dev/null'; %s. Exiting.",
			g_strerror(errno));
		mce_log_close();
		exit(EXIT_FAILURE);
	}

	if ((dup(i) == -1)) {
		mce_log(LL_CRIT,
			"Failed to dup() `/dev/null'; %s. Exiting.",
			g_strerror(errno));
		mce_log_close();
		exit(EXIT_FAILURE);
	}

	/* Set umask */
	umask(022);

	/* Set working directory */
	if ((chdir("/tmp") == -1)) {
		mce_log(LL_CRIT,
			"Failed to chdir() to `/tmp'; %s. Exiting.",
			g_strerror(errno));
		mce_log_close();
		exit(EXIT_FAILURE);
	}

	/* Single instance */
	if ((i = open(MCE_LOCKFILE, O_RDWR | O_CREAT, 0640)) == -1) {
		mce_log(LL_CRIT,
			"Cannot open lockfile; %s. Exiting.",
			g_strerror(errno));
		mce_log_close();
		exit(EXIT_FAILURE);
	}

	if (lockf(i, F_TLOCK, 0) == -1) {
		mce_log(LL_CRIT, "Already running. Exiting.");
		mce_log_close();
		exit(EXIT_FAILURE);
	}

	sprintf(str, "%d\n", getpid());
	write(i, str, strlen(str));
	close(i);

	/* Ignore TTY signals */
	signal(SIGTSTP, SIG_IGN);
	signal(SIGTTOU, SIG_IGN);
	signal(SIGTTIN, SIG_IGN);

	/* Ignore child terminate signal */
	signal(SIGCHLD, SIG_IGN);

EXIT:
	return 0;
}

/**
 * Main
 *
 * @param argc Number of command line arguments
 * @param argv Array with command line arguments
 * @return 0 on success, non-zero on failure
 */
int main(int argc, char **argv)
{
	int optc;
	int opt_index;

	int verbosity = LL_DEFAULT;
	int logtype = -1;

	gint status = 0;
	gboolean show_module_info = FALSE;
	gboolean daemonflag = FALSE;
	gboolean systembus = TRUE;
	gboolean debugmode = FALSE;

	const char optline[] = "dS";

	struct option const options[] = {
		{ "daemonflag", no_argument, 0, 'd' },
		{ "force-syslog", no_argument, 0, 's' },
		{ "force-stderr", no_argument, 0, 'T' },
		{ "session", no_argument, 0, 'S' },
		{ "show-module-info", no_argument, 0, 'M' },
		{ "debug-mode", no_argument, 0, 'D' },
		{ "quiet", no_argument, 0, 'q' },
		{ "verbose", no_argument, 0, 'v' },
		{ "help", no_argument, 0, 'h' },
		{ "version", no_argument, 0, 'V' },
		{ 0, 0, 0, 0 }
	};

	/* NULL the mainloop */
	mainloop = NULL;

	/* Initialise support for locales, and set the program-name */
	if (init_locales(PRG_NAME) != 0)
		goto EXIT;

	/* Parse the command-line options */
	while ((optc = getopt_long(argc, argv, optline,
				   options, &opt_index)) != -1) {
		switch (optc) {
		case 'd':
			daemonflag = TRUE;
			break;

		case 's':
			if (logtype != -1) {
				usage();
				status = EINVAL;
				goto EXIT;
			}

			logtype = MCE_LOG_SYSLOG;
			break;

		case 'T':
			if (logtype != -1) {
				usage();
				status = EINVAL;
				goto EXIT;
			}

			logtype = MCE_LOG_STDERR;
			break;

		case 'S':
			systembus = FALSE;
			break;

		case 'M':
			show_module_info = TRUE;
			break;

		case 'D':
			debugmode = TRUE;
			break;

		case 'q':
			if (verbosity > LL_NONE)
				verbosity--;
			break;

		case 'v':
			if (verbosity < LL_DEBUG)
				verbosity++;
			break;

		case 'h':
			usage();
			goto EXIT;

		case 'V':
			version();
			goto EXIT;

		default:
			usage();
			status = EINVAL;
			goto EXIT;
		}
	}

	/* We don't take any non-flag arguments */
	if ((argc - optind) > 0) {
		fprintf(stderr,
			_("%s: Too many arguments\n"
			  "Try: `%s --help' for more information.\n"),
			progname, progname);
		status = EINVAL;
		goto EXIT;
	}

	if (logtype == -1)
		logtype = (daemonflag == TRUE) ? MCE_LOG_SYSLOG :
						 MCE_LOG_STDERR;

	mce_log_open(PRG_NAME, LOG_DAEMON, logtype);
	mce_log_set_verbosity(verbosity);

	/* Daemonize if requested */
	if (daemonflag == TRUE)
		daemonize();

	signal(SIGUSR1, signal_handler);
	signal(SIGHUP, signal_handler);
	signal(SIGTERM, signal_handler);

	/* Initialise GType system */
	g_type_init();

	/* Register a mainloop */
	mainloop = g_main_loop_new(NULL, FALSE);

	/* Initialise subsystems */

	/* Get configuration options */
	/* ignore errors; this way the defaults will be used if
	 * the configuration file is invalid or unavailable
	 */
	(void)mce_conf_init();

	/* Initialise D-Bus */
	if (mce_dbus_init(systembus) == FALSE) {
		mce_log(LL_CRIT,
			"Failed to initialise D-Bus");
		mce_log_close();
		exit(EXIT_FAILURE);
	}

	/* Initialise GConf
	 * pre-requisite: g_type_init()
	 */
	if (mce_gconf_init() == FALSE) {
		mce_log(LL_CRIT,
			"Cannot connect to default GConf engine");
		mce_log_close();
		exit(EXIT_FAILURE);
	}

	/* Setup all datapipes */
	setup_datapipe(&system_state_pipe, READ_WRITE, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(MCE_STATE_UNDEF));
	setup_datapipe(&master_radio_pipe, READ_WRITE, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(0));
	setup_datapipe(&call_state_pipe, READ_WRITE, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(CALL_STATE_NONE));
	setup_datapipe(&call_type_pipe, READ_WRITE, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(NORMAL_CALL));
	setup_datapipe(&alarm_ui_state_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(MCE_ALARM_UI_INVALID_INT32));
	setup_datapipe(&submode_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(MCE_NORMAL_SUBMODE));
	setup_datapipe(&display_state_pipe, READ_WRITE, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(MCE_DISPLAY_UNDEF));
	setup_datapipe(&display_brightness_pipe, READ_WRITE, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(0));
	setup_datapipe(&led_brightness_pipe, READ_WRITE, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(0));
	setup_datapipe(&led_pattern_activate_pipe, READ_ONLY, FREE_CACHE,
		       0, NULL);
	setup_datapipe(&led_pattern_deactivate_pipe, READ_ONLY, FREE_CACHE,
		       0, NULL);
	setup_datapipe(&key_backlight_pipe, READ_WRITE, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(0));
	setup_datapipe(&keypress_pipe, READ_ONLY, FREE_CACHE,
		       sizeof (struct input_event), NULL);
	setup_datapipe(&touchscreen_pipe, READ_ONLY, FREE_CACHE,
		       sizeof (struct input_event), NULL);
	setup_datapipe(&device_inactive_pipe, READ_WRITE, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(FALSE));
	setup_datapipe(&lockkey_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(0));
	setup_datapipe(&keyboard_slide_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(0));
	setup_datapipe(&lid_cover_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(0));
	setup_datapipe(&lens_cover_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(0));
	setup_datapipe(&proximity_sensor_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(0));
	setup_datapipe(&tk_lock_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(LOCK_UNDEF));
	setup_datapipe(&charger_state_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(0));
	setup_datapipe(&battery_status_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(BATTERY_STATUS_UNDEF));
	setup_datapipe(&battery_level_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(100));
	setup_datapipe(&camera_button_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(CAMERA_BUTTON_UNDEF));
	setup_datapipe(&inactivity_timeout_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(DEFAULT_INACTIVITY_TIMEOUT));
	setup_datapipe(&audio_route_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(AUDIO_ROUTE_UNDEF));
	setup_datapipe(&usb_cable_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(0));
	setup_datapipe(&jack_sense_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(0));
	setup_datapipe(&power_saving_mode_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(0));
	setup_datapipe(&thermal_state_pipe, READ_ONLY, DONT_FREE_CACHE,
		       0, GINT_TO_POINTER(THERMAL_STATE_UNDEF));

	/* Initialise mode management
	 * pre-requisite: mce_gconf_init()
	 * pre-requisite: mce_dbus_init()
	 */
	if (mce_mode_init() == FALSE) {
		status = EXIT_FAILURE;
		goto EXIT;
	}

	/* Initialise DSME
	 * pre-requisite: mce_gconf_init()
	 * pre-requisite: mce_dbus_init()
	 * pre-requisite: mce_mce_init()
	 */
	if (mce_dsme_init(debugmode) == FALSE) {
		if (debugmode == FALSE) {
			mce_log(LL_CRIT, "Cannot connect to DSME");
			status = EXIT_FAILURE;
			goto EXIT;
		}
	}

	/* Initialise powerkey driver */
	if (mce_powerkey_init() == FALSE) {
		status = EXIT_FAILURE;
		goto EXIT;
	}

	/* Initialise /dev/input driver
	 * pre-requisite: g_type_init()
	 */
	if (mce_input_init() == FALSE) {
		status = EXIT_FAILURE;
		goto EXIT;
	}

	/* Initialise switch driver */
	if (mce_switches_init() == FALSE) {
		status = EXIT_FAILURE;
		goto EXIT;
	}

	/* Initialise tklock driver */
	if (mce_tklock_init() == FALSE) {
		status = EXIT_FAILURE;
		goto EXIT;
	}

	/* Load all modules */
	if (mce_modules_init() == FALSE) {
		status = EXIT_FAILURE;
		goto EXIT;
	}

	if (show_module_info == TRUE) {
		mce_modules_dump_info();
		goto EXIT;
	}

	/* Run the main loop */
	g_main_loop_run(mainloop);

	/* If we get here, the main loop has terminated;
	 * either because we requested or because of an error
	 */
EXIT:
	/* Unload all modules */
	mce_modules_exit();

	/* Call the exit function for all components */
	mce_tklock_exit();
	mce_switches_exit();
	mce_input_exit();
	mce_powerkey_exit();
	mce_dsme_exit();
	mce_mode_exit();

	/* Free all datapipes */
	free_datapipe(&thermal_state_pipe);
	free_datapipe(&power_saving_mode_pipe);
	free_datapipe(&jack_sense_pipe);
	free_datapipe(&usb_cable_pipe);
	free_datapipe(&audio_route_pipe);
	free_datapipe(&inactivity_timeout_pipe);
	free_datapipe(&battery_level_pipe);
	free_datapipe(&battery_status_pipe);
	free_datapipe(&charger_state_pipe);
	free_datapipe(&tk_lock_pipe);
	free_datapipe(&proximity_sensor_pipe);
	free_datapipe(&lens_cover_pipe);
	free_datapipe(&lid_cover_pipe);
	free_datapipe(&keyboard_slide_pipe);
	free_datapipe(&lockkey_pipe);
	free_datapipe(&device_inactive_pipe);
	free_datapipe(&touchscreen_pipe);
	free_datapipe(&keypress_pipe);
	free_datapipe(&key_backlight_pipe);
	free_datapipe(&led_pattern_deactivate_pipe);
	free_datapipe(&led_pattern_activate_pipe);
	free_datapipe(&led_brightness_pipe);
	free_datapipe(&display_brightness_pipe);
	free_datapipe(&display_state_pipe);
	free_datapipe(&submode_pipe);
	free_datapipe(&alarm_ui_state_pipe);
	free_datapipe(&call_type_pipe);
	free_datapipe(&call_state_pipe);
	free_datapipe(&master_radio_pipe);
	free_datapipe(&system_state_pipe);

	/* Call the exit function for all subsystems */
	mce_gconf_exit();
	mce_dbus_exit();
	mce_conf_exit();

	/* If the mainloop is initialised, unreference it */
	if (mainloop != NULL)
		g_main_loop_unref(mainloop);

	/* Log a farewell message and close the log */
	mce_log(LL_INFO, "Exiting...");
	mce_log_close();

	return status;
}
